compiler/rustc_passes/src/dead.rs

```diff
@@ -666,7 +679,34 @@ impl<'tcx> DeadVisitor<'tcx> {
             self.tcx.struct_span_lint_hir(lint::builtin::DEAD_CODE, id, span, |lint| {
                 let def_id = self.tcx.hir().local_def_id(id);
                 let descr = self.tcx.def_kind(def_id).descr(def_id.to_def_id());
-                lint.build(&format!("{} is never {}: `{}`", descr, participle, name)).emit()
+                let mut err = lint.build(&format!("{} is never {}: `{}`", descr, participle, name));
+                let hir = self.tcx.hir();
+                if let Some(encl_scope) = hir.get_enclosing_scope(id) {
+                    if let Some(encl_def_id) = hir.opt_local_def_id(encl_scope) {
+                        if let Some(ign_traits) = self.ignored_derived_traits.get(&encl_def_id.to_def_id()) {
+                            let traits_str = ign_traits
+                                .iter()
+                                .map(|t| format!("`{}`", self.tcx.item_name(*t))).collect::<Vec<_>>()
+                                .join(" and ");
+                            let plural_s = pluralize!(ign_traits.len());
+                            let article = if ign_traits.len() > 1 { "" } else { "a " };
+                            let is_are = if ign_traits.len() > 1 { "these are" } else { "this is" };
+                            let msg = format!("`{}` has {}derived impl{} for the trait{} {}, but {} ignored during dead code analysis",
+                                              self.tcx.item_name(encl_def_id.to_def_id()),
+                                              article,
+                                              plural_s,
+                                              plural_s,
+                                              traits_str,
+                                              is_are);
+                            if let Some(span) = self.tcx.def_ident_span(encl_def_id) {
+                                err.span_note(span, &msg);
```

Comment:
would it be possible to get the span of the `Debug` impl instead?

Reply:
Done, the note now points to the part of the derive macro invocation where the `Clone`/`Debug` impl comes from. I've also added the word "intentionally" to the note message for clarity.